generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/internal/types/generateDirective.kt

```diff
@@ -66,43 +69,51 @@ internal fun generateEnumValueDirectives(generator: SchemaGenerator, field: Fiel
-            .name(directiveInfo.effectiveName)
-            .description(directiveInfo.directiveAnnotation.description)
-            .repeatable(directiveInfo.repeatable)
+    val directive = generator.directives.computeIfAbsent(directiveName) {
+        val directiveFromHook = generator.config.hooks.willGenerateDirective(directiveInfo)
+        val generatedDirective = if (directiveFromHook != null) {
+            directiveFromHook
+        } else {
+            val builder = GraphQLDirective.newDirective()
+                .name(directiveInfo.effectiveName)
+                .description(directiveInfo.directiveAnnotation.description)
+                .repeatable(directiveInfo.repeatable)
 
-        directiveInfo.directiveAnnotation.locations.forEach {
-            builder.validLocation(it)
-        }
+            directiveInfo.directiveAnnotation.locations.forEach {
+                builder.validLocation(it)
+            }
 
-        val directiveClass: KClass<out Annotation> = directiveInfo.directive.annotationClass
-        val directiveArguments: List<KProperty<*>> = directiveClass.getValidProperties(generator.config.hooks)
+            val directiveClass: KClass<out Annotation> = directiveInfo.directive.annotationClass
+            val directiveArguments: List<KProperty<*>> = directiveClass.getValidProperties(generator.config.hooks)
 
-        directiveArguments.forEach { prop ->
-            val argument = generateDirectiveArgument(prop, generator)
-            builder.argument(argument)
+            directiveArguments.forEach { prop ->
+                val argument = generateDirectiveArgument(prop, generator)
+                builder.argument(argument)
+            }
+            builder.build()
         }
-
-        builder.build()
+        generator.config.hooks.didGenerateDirective(directiveInfo, generatedDirective)
     }
 
     val appliedDirective = directive.toAppliedDirective()
     return if (directive.arguments.isNotEmpty()) {
         appliedDirective.transform { builder ->
             directiveInfo.directive.annotationClass.getValidProperties(generator.config.hooks).forEach { prop ->
-                directive.getArgument(prop.name)
-                    ?.toAppliedArgument()
-                    ?.transform { argumentBuilder ->
-                        val value = prop.call(directiveInfo.directive)
-                        argumentBuilder.valueProgrammatic(value)
-                    }
-                    ?.let { appliedDirectiveArgument ->
-                        builder.argument(appliedDirectiveArgument)
+                val argumentToBeModified = directive.getArgument(prop.name)
+                if (argumentToBeModified != null) {
+                    val value = prop.call(directiveInfo.directive)
+                    // annotations cannot have null values, we should skip known defaults
+                    if (value != DEFAULT_DIRECTIVE_STRING_VALUE && value != DEFAULT_DIRECTIVE_INT_VALUE) {
```

Comment:
Currently we don't have mechanism for skipping the default (i.e., not provided) values. Since we cannot use `null` in annotations we should have some explicit (unlikely) defaults.